### ISSUE TITLE
Non-type-use nullability annotations should not be reinterpreted as type-use

### ISSUE THREAD

**Issue by astiob** [typetools/checker-framework#1667]
Nullness Checker currently treats various legacy nullness annotations, e. g. `@javax.annotation.CheckForNull`, as full aliases to `@org.checkerframework.checker.nullness.qual.Nullable` etc. However, most if not all of them (certainly at least the `javax.annotation` ones) are not `@Target(ElementType.TYPE_USE)` annotations. This means that the following code is treated differently by Checker Framework and by other tools:

```java
import javax.annotation.CheckForNull;
class Test {
    @CheckForNull Object[] foo(@CheckForNull Object[] bar) {
    	return null;
    }
    void test() {
    	foo(null);
    }
}
```

The traditional meaning of this code, supported by the `jsr305` Javadoc (quoted below), and surely assumed by all programmers and validation/bug-seeking tools except Checker Framework’s Nullness Checker, is that the parameter (`bar`, which has type `Object[]`) can be null and the return value (which also has type `Object[]`) can also be null. Nothing is said about the contents (elements) of the parameter and of the return value. Quoting the [`jsr305` Javadoc](http://static.javadoc.io/com.google.code.findbugs/jsr305/3.0.2/javax/annotation/CheckForNull.html):

> The annotated element might be null, and uses of the element should check for null.
>
> When this annotation is applied to a method it applies to the method return value.

This annotation has no explicit `@Target` meta-annotation, so the “annotated element” here is the parameter (`Object[] bar`) and the method.

Checker Framework, on the other hand, pretends that this is a type-use annotation and, per the JSR-308 parsing rules, applies it to the nearest element of the type, which is `Object`, the type of the array element. So it uses the default qualifier for the array itself, which is `@NonNull`, and fails to compile this class.

**Checker Framework changes the meaning of this code.** Not only is this inconvenient, it is also dangerous.

[Checker Framework’s documentation](https://checkerframework.org/manual/#nullness-related-work) even makes the following statement on this matter:

> Note that some older tools interpret array and varargs declarations inconsistently with the Java specification. For example, they might interpret `@NonNull Object []` as “non-null array of objects”, rather than as “array of non-null objects” which is the correct Java interpretation. Such an interpretation is unfortunate and confusing. See Section 31.6.3 for some more details about this issue.

This is simply incorrect. Older tools interpret their annotations in perfect accordance with how non-`TYPE_USE` annotations are to be interpreted: they apply to the whole declaration of the field, parameter or method. I think it is fair to say that if someone says their field can be null, then they mean just that, regardless of whether the field is a reference to an array or to something else. (In fact, for something like `@CheckForNull java.lang.Object[] bar`, the type-use interpretation isn’t even possible, as this would be invalid syntax if `@CheckForNull` was [only] `TYPE_USE`, much less correct.)

**Checker Framework should detect when annotations apply to field, parameter and method declarations and move them to the top-level type of the declared field, parameter or return value** (the last component of the type in Java syntax).

Keep in mind that some annotations might, unfortunately, have both `TYPE_USE` and `FIELD` etc. at the same time. A quick test tells me that in case of ambiguity the annotation applies *both* to the type and to the field. In practice, this is used by [Bean Validation 2.0](http://beanvalidation.org/2.0/spec/), and unfortunately, the meaning they assign to it is weirder than that. As far as I can tell, they _ignore_ type-use annotations anywhere but in generic type arguments, e. g. `List<@NotNull String>`. Thus, even if there is an ambiguous situation such as `@NotNull String[] field`, the field annotation applies only to the whole field (the array here) and the type-use instance is ignored. There are also cases when a Bean Validation 2.0 annotation is unambiguously type-use and is ignored entirely, e. g. `String @NotNull[]`. Checker Framework, of course, currently doesn’t ignore them. This may be viewed as a defect in Bean Validation (assuming my understanding of it is even right), and I doubt that replicating this in Checker Framework is worth the effort, although from a programmer’s perspective, it would be great if Checker Framework could detect this and warn the programmer during compilation.

**Comment by mernst** [comment]
Thanks for the suggestion.

The Checker Framework should issue a warning to warn users of the difference in semantics.
We'll put that on our to-do list, but if you want to submit a pull request before then, that would be great and would surely mean that the feature would be implemented sooner.

<br>

One can imagine two different purposes for supporting older tools' annotations.

 * To emulate other tools' behavior.
 * To ease users in the transition to using the Checker Framework.

It sounds like you want the former, but that is not the aim of the Checker Framework.  If you want a tool that treats `@NonNull` as a declaration annotation, or that has other buggy behavior, then you should use one, but the Checker Framework is not that tool. :-)

You are proposing that the declaration `@NonNull Object[]` would mean different things depending on an import statement at the top of a file.  This would violate a design aim.  Users should be able to use the Checker Framework's support for other tools' annotations until they change their import statements, and the Checker Framework's behavior should not depend on the `import` statements.

<br>

I called treating `@NonNull` as a declaration annotation buggy, and that deserves some elaboration.

`@NonNull` is a restriction on values, not a fact about a variable or a method.
Therefore, `@NonNull` should be a type annotation, not a field or method annotation.

Many tools and IDEs have recognized that `@NonNull` should be a type annotation and changed their definitions accordingly.  FindBugs hasn't made this change, but maybe that's because FindBugs has been abandoned by its maintainer.  Even if programmers made an incorrect assumption, that would be an argument to educate programmers, not to change the Checker Framework's semantics or to emulate less capable tools.

> The traditional meaning of this code, supported by the jsr305 Javadoc (quoted below), and surely assumed by all programmers and validation/bug-seeking tools except Checker Framework’s Nullness Checker

First, the type annotation interpretation is the standard and correct meaning today, supported by many tools even if some tools lag behind.
Second, JSR 305 is not the "traditional meaning".  JSR 305 was a misguided attempt to formalize FindBugs's buggy implementation.  The flaws in JSR 305 were recognized even at the time, and JSR 305 was abandoned in 2009 as a result.  Quoting the JSR 305 Javadoc is quoting a known mistake and does not support your argument. :-)

<br>

As I noted above, even if we don't want to change the semantics of the annotations, the Checker Framework should issue a warning.
The warning should be issued only when a legacy declaration annotation is written on a field of array type or a method whose return type is an array type.
Here are two possible implementation strategies:
 * Hard-code the list of such annotations.
 * Change all the cleanroom annotation implementations to be consistent with their original implementations, rely on the Checker Framework's existing support for moving annotations from declaration to type locations, and look up each annotation's `@Target` meta-annotation.

The first approach would be quicker to complete, but the second would be more general.

**Comment by astiob** [comment]
Thanks for your reply. I apologize in advance if I come across as slightly aggressive at any point, but I do feel quite strongly about this :-) But I imagine you do as well. I feel we’re still not quite on the same page, and I’m hoping that I’ll be able to explain my concerns properly.

I challenge you (I will so regret this…) to find a single project or developer that uses `@javax.annotation.Nullable Object[] field;` and actually means that the array elements can be null but the array cannot. I’m confident that there are many projects (especially old ones) and developers that use the same to mean that the array itself cannot be null.

You are right that some other tools have switched from declaration annotations to type-use annotations. This came with huge backwards-compatibility warnings and guides instructing users to change their code to move the annotations to the newly-appropriate places. The users’ old code was perfectly well-meaning and correctly checked by the old tools, and their new code is perfectly well-meaning and correctly checked by the new tools. (This suggests another reason not to include third-party annotations in Checker Framework or somehow deprioritize them on the classpath, so that if code using e. g. Eclipse’s annotations includes them at runtime, CF can check whether they’re meant to be type-use or not.)

What Checker Framework does here is it takes the old code and silently reinterprets it according to the new, incompatible definition. Again, **it changes the meaning of the code.** CF loves soundness, but could anything be less sound than the checker moving type qualifiers at whim?

~I realize that there may be cases where this is unavoidable: for example, if a file uses Eclipse’s annotations but they aren’t on the classpath, so it’s impossible to discern whether the author used the old or the new definitions of the annotations.~ …Actually, wait. Is this possible? If this is a class file, then it’s clear which annotations are type-use and which are for declarations because the compiler took care of that; and if this is a source file, then the compiler (even without CF) will refuse to compile the file until the annotation definitions are added to the classpath. So actually, it is always unambiguously clear which annotations apply to what.

I can’t imagine anyone intentionally using Checker Framework with FindBugs annotations and wanting them to work like Checker Framework’s native ones. How is this easier than importing CF’s annotations instead?

On the other hand, there are third-party libraries that already include nullness annotations that people want to use. In my case, this is Spring 5. I want to use Spring and I want to use Checker Framework in my code. With Spring 4, I had to write stubs for any Spring code I used to show CF where nulls were allowed. With Spring 5 and the fix for #1664, Checker Framework thinks it already knows this—and I would think so too, after seeing in the manual that CF claims to support Spring’s annotation. But it actually **breaks my code** by assuming things about Spring that Spring never promised. So I will have to continue writing stubs to fix the instances where Spring uses arrays. This may be fine for me, but others won’t know this until they get hit by a runtime exception that CF guaranteed to avoid. (Then again, stubs will still be necessary for any instances of nullable array elements.)

> You are proposing that the declaration `@NonNull Object[]` would mean different things depending on an import statement at the top of a file.

Yes.

> This would violate a design aim. Users should be able to use the Checker Framework's support for other tools' annotations until they change their import statements, and the Checker Framework's behavior should not depend on the import statements.

(Oh, you use “the” with “Checker Framework”. Is this the preferred way?)

Users should be able to use the Checker Framework’s support for other tools’ annotations until they change their code to use Checker Framework’s annotations (yes, change their import statements, but it’s more than that). The behaviour should stay the same as they had with their previous tool until then. Is this not the best way to ease the transition to the Checker Framework, just as you want? When the user changes the import statement in this file, they’ll go through all annotation instances in the same file and fix them up.

If CF’s only reason to support FindBugs annotations is to avoid changing import statements, it would be better to drop the support entirely than to delude end users into believing that their code is still being correctly checked. Changing import statements in this case is a trivial find-and-replace and can be done at the same time as the Checker Framework is introduced into the project. This will alert the user, however, that the meaning of their annotations may have changed and they need to double-check that everything is proper, and they will not get the false sense that third-party libraries with FindBugs annotations are correctly checked either.

> First, the type annotation interpretation is the standard and correct meaning today, supported by many tools even if some tools lag behind.

Only for annotations that are explicitly declared `TYPE_USE`. And it should stay that way.

> Second, JSR 305 is not the "traditional meaning". JSR 305 was a misguided attempt to formalize FindBugs's buggy implementation. The flaws in JSR 305 were recognized even at the time, and JSR 305 was abandoned in 2009 as a result. Quoting the JSR 305 Javadoc is quoting a known mistake and does not support your argument. :-)

I never intended to discuss whether JSR-305 or FindBugs were good or bad. But I can’t agree that they’re “buggy” when they do exactly what it says on the tin; and if the meaning that has been universally used for many years until the Checker Framework came along, and is still used for those same annotations, is not “traditional”, then I don’t know what it is. Finally, I’m quoting the Javadoc of the thing the Javadoc defines because, well, it defines the thing, and CF violates that definition.

I have absolutely no objections to the Checker Framework’s type-use annotations—I love them! (Although I do find it unfortunate that this syntax ambiguity was allowed to arise.) I completely agree that without this, the type system is incomplete and more or less useless. This is why I chose to use the Checker Framework in the first place!

But CF shouldn’t claim to support FindBugs annotations and then assign an alien meaning to them that contradicts the intentions of the people who added them to their code.

---

A warning seems good regardless, as e. g. `@javax.annotation.Nullable Object[]` lacks full type information in any case: if this is read as a declaration annotation, this says nothing about the nullability of individual elements (it doesn’t allow it, but it doesn’t really allow it either—how could it, if `javax.annotation.Nullable` is not `TYPE_USE`?).

I can see a little sense in warning users of old Eclipse annotations and pretending CF is a new Eclipse and hence applies the new, type-use definitions of the annotations. However, if the user has the old definitions on their classpath, then it is very clear what their annotations mean, and CF should follow that definition to check the code the best that it can. (And if they have the new definitions on their classpath, then everyone is already happy and no warnings or reinterpretation is needed.)

However, for annotations such as those of FindBugs, there is only one definition, and these annotations are always, always declaration annotations. CF should treat them as such regardless of whether it warns about it.

All in all, when compiling source code, indeed the two options arise that you suggested:

> * Hard-code the list of such annotations.
> * Change all the cleanroom annotation implementations to be consistent with their original implementations, rely on the Checker Framework's existing support for moving annotations from declaration to type locations, and look up each annotation's @Target meta-annotation.
>
> The first approach would be quicker to complete, but the second would be more general.

I completely agree on all points.

I would add that the Checker Framework’s cleanroom implementations should be omitted from the classpath during lookups arising from direct mentions of annotations in source files, at least the ones like Eclipse’s that have historically changed meaning, so that user code that mentions an Eclipse `@Nullable Object[]` rather fails to compile if the user forgets to import the annotation on the classpath than is mischecked if the user copied this from an old project that used declaration annotations.

For classes that are already in compiled form during compilation (external library dependencies), the information on which annotations are type-use and which are declaration annotations is already baked into the class files, as far as I know. CF should honour this information, regardless of which annotation classes are being used. I’m guessing that at the moment CF explicitly moves any annotations it detects as declaration annotations in class files to the innermost type-use level. It is my humble guess that changing this to move them to the outermost type-use level should be easy. CF should warn when (and only when) an array field, parameter or method return value is used that is annotated with a declaration annotation. Ideally, it should not warn when an external library contains such a declaration that is never used by user code, but I don’t know how hard this would be to implement.

**Comment by smillst** [comment]
`@javax.annotation.Nullable Object[]` is now treated as a possibly null array of non-null objects.  See 13f65f8 for more details.

**Comment by astiob** [comment]
Awesome, thanks!
